### DISCUSSION TITLE
Why does the RSpec/SubjectStub cop exist?

### DISCUSSION THREAD

**Discussion by DannyBen** [rubocop-hq/rubocop-rspec#1492] [Q&A]
Can anyone help me understand the logic behind the `RSpec/SubjectStub` cop?

It disallows stubbing of inner methods of the subject under test, but to me this seems like a legitimate, intuitive and readable approach. I do not see the downside, and any alternative I can think of would require more coding and less readability.

Consider this test subject:

```ruby
# stock.rb
class Stock
  def price
    raise "Invalid data" unless valid?
    12.3
  end

  def valid?
    # some logic
    true
  end
end

```

and this spec:

```ruby
# stock_spec.rb
require './stock'

describe Stock do
  subject(:stock) { described_class.new }

  describe '#price' do
    it 'returns the price' do
      expect(stock.price).to eq 12.3
    end

    context 'when it is invalid' do
      before { allow(stock).to receive(:valid?).and_return false }

      it 'raises an error' do
        expect { stock.price }.to raise_error('Invalid data')
      end
    end
  end
end
```

running `rubocop --only RSpec` will trigger this offense:

```
RSpec/SubjectStub: Do not stub methods of the object under test
```

why is this bad? and what is an acceptable alternative?

**Comment by pirj** [comment]
It exists to back [this guideline](https://rspec.rubystyle.guide/#dont-stub-subject), originated [here](https://github.com/rubocop/rspec-style-guide/issues/32).

Speaking of the example itself, I see an option to avoid stubbing by passing invalid values to the object initializer (e.g. `new(-1)`). This, though, would increase the testing surface to `valid?`, that is most probably covered in other examples.
On the other hand, this would prevent from relying on the implementation details.

**Reply by DannyBen** [reply]
I have seen these and similar statements about it being a code smell, but none with any convincing explanation of why.

I know there are multiple ways to avoid stubbing it, I am just not convinced there is any need to.
Also, of course `valid?` is just an example, and my tests are stubbing `subject` methods in many projects without any side effect or any of them feeling "smelly".

Imagine a `#market_open?` method on the same object - what is more natural than to `expect(subject).to receive(:market_open).and_return(false)` - it's exactly the same sentence if you write it in code, and say it in English. A win in my book.

Also, this way is universal and works with almost any use case. No need to think if there is an object that I need to mock, or some other acrobatics to avoid this cop.

I guess this will have to be just one of those cops that I always disable.

**Reply by pirj** [reply]
I hope you'll find explanations from [Effective Testing with RSpec](https://medium.com/pragmatic-programmers/stubject-stubbing-the-subject-84a7df730027), [Thoughtbot](https://robots.thoughtbot.com/don-t-stub-the-system-under-test) and [a former RSpec maintainer](https://samphippen.com/introducing-rspec-smells-and-where-to-find-them#smell-1-stubject) convincing enough.

For some reason, the [links with `@see` from the cop's code](https://github.com/rubocop/rubocop-rspec/blob/0d3b2a5d0351fa1c6230930849f744b408078ff6/lib/rubocop/cop/rspec/subject_stub.rb#L13) don't appear in [the docs](https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#rspecsubjectstub).

**Reply by DannyBen** [reply]
Thanks for engaging, and for the links.
The `@see` links are definitely helpful. Two key argument against stubbing other subject methods that I found on [this page](https://penelope.zone/2015/12/27/introducing-rspec-smells-and-where-to-find-them.html#smell-1-stubjec) are this:

> That modification means that we aren’t testing the actual class of that object as it will go into production. Instead, when we stub an object, we’re creating a new class (literally) that only exists in the testing environment.
> ...
> This pattern is problematic because it means that the implementation being tested is one that will never see production

While I totally agree that whenever possible, and whenever natural to do so, the object boundaries should be redefined to allow testing "the full object" as it is. However, I am still not sure I agree that this is always needed. 

My primary argument is this: When I am testing a single method of an object, I test just that. The method. Not the object.
I feel free to stub and fake anything other than the **method** under test. This of course, comes with the obligation to ensure that I also have reasonable tests for that *other* methods that may be stubbed.

```ruby
describe '#price' do
  context "when the market is closed" do
    before { allow(subject).to receive(:market_open?).and_return false }

    it { ... }
    it { ... }
  end
end

describe '#market_open?' do
  context "when it is Sunday" do
    before { allow(Date).to receive(...).and_return ... }

    it { is_expected.to be false }
  end
end
```

Of course this example can be written differently, but it is just one example use case our of many different ones that have the same solution that requires no overthinking, no over-engineering and so far for me, was quite maintainable and readable even when I come back to code written years ago.

What I think I will do is let this cop *remind me* to reconsider any subject stubbing, but at this point I don't feel I need to comply with it at all costs.

**Reply by pirj** [reply]
> the object boundaries should be redefined to allow testing "the full object" as it is
> When I am testing a single method of an object, I test just that. The method. Not the object.

I agree with you.

The cop acts very formally. It checks if `subject`'s methods are stubbed.
If the `subject` was defined as `subject(:price) { market.price }`, as you say you consider the method a subject under test, and you would stub other methods, subject's siblings:
```ruby
before { allow(market).to receive(:market_open?).and_return false }
```
the cop would not trigger.

**Reply by DannyBen** [reply]
> if the subject was defined as subject(:price) { market.price } ... the cop would not trigger.

That is a great point!

It is brings up another little confusion I have with the cops and with designing specs in general.
I never used to create subjects for each method. To me, the subject was always the object under test. It made me scratch my head when I saw the cop that requires named subjects. When I started naming my subjects, the tests looked less readable since instead of `expect(subject).to....` I had an `expect(some_name).to...` which made the subject indistinguishable from other variables.

Having the subject be the method (`subject { stock.price }`) made sense to me, when using the one liner syntax:

```ruby
it { is_expected.to eq stock.last_price }
```

but then I realized that the one liner syntax is discouraged.... so it discourages me from using named (or even multiple) subjects.

I guess my point is, rspec - due to its age - has multiple "generations" of accepted / best practice approaches, and it is sometimes conflicting and confusing to figure out which is best (another example would be spies / doubles and the preference for `have_received` over `receive`).

A common pattern I use is to define one subject "construction pattern" and then only change parameters in contexts:

```ruby
describe Stock do
  subject { described_class.new id }
  let(:id) { :GOOG }

  it { ... }

  context "with an ETF" do
    let(:id) { :SPY }

    it { ... }
  end
end
```

(and when using this approach, which I love, sometimes triggers another cop - that complains I have too many `let` statements... even when I have 3 or 4 only).

**Reply by pirj** [reply]
> one liner syntax is discouraged

Is it?

> subject "construction pattern"

Another way that I believe Effective Testing with RSpec suggests is to use methods building the subject:
```ruby
def stock_for(id)
  described_class.new(id)
end

it { expect(stock_for(:SPY)).to ... }
```

But with more than 3-4 inputs to the constructor, it can quickly get out of control, too.

**Reply by DannyBen** [reply]
> is it?

From what I see, absolutely.
I could not find the posts where I actually saw it first, but a quick google search shows some links where it is discussed negatively:

- https://github.com/rubocop/rspec-style-guide/issues/76
- https://blog.rubystyle.guide/rspec/2019/07/17/rspec-implicit-block-syntax.html
- https://www.reddit.com/r/ruby/comments/cejl3q/call_for_discussion_rspec_implicit_block/?st=jy8s65z0&sh=0e5bb1da
- https://github.com/rspec/rspec-expectations/issues/805

Notice this comment in particular: https://github.com/rspec/rspec-expectations/issues/805#issuecomment-112238533

> Q: Is one-liner syntax something that we (as a core team) use much, or is it in the `expect_any_instance_of` bucket of extras?
> A: It's in the same bucket, I certainly don't recommend it.

Furthermore, I am using block expectations quite a lot for output and error checking using my [rspec_approvals](https://github.com/DannyBen/rspec_approvals) interactive approval matchers, and `is_expected` cannot be used with block matchers, which will mean I will end up with a mixed bag of rspec syntaxes.

Another hint, is that the `its(method) { is_expected.to ... }` was removed.

I wish they were both alive, but I can understand the reasoning.

> Another way that I believe Effective Testing with RSpec suggests is to use methods building the subject:

The top appealing rspec feature for me, is that it is a readable DSL.
It is a priority for me that the specs are readable, and easy to write, and utilize the DSL aspects that make it read like English.
Otherwise, I would have used assertions with minitest or something... :)

I want: `expect(something).to ...` with as little fluff in between as possible.

I avoid `def` in specs at all costs, and for the same reason I even avoid factories and complicated stubbing. So far this approach has served me quite well - I feel my tests are efficient and effective, but maybe I am just used to them.

**Reply by pirj** [reply]
> one liner syntax is discouraged

Those links are all about the implicit block expectation syntax, not specifically one-liner expectation syntax.
The fact that the core team doesn't use it doesn't mean that everyone else should not, too.
As far as I'm aware, `is_expected` is not going anywhere.


> `its(method) { is_expected.to ... }` was removed

[Was it](https://github.com/rspec/rspec-its/blob/86de451da52aae1a4d0d057de05fb23627d478d7/features/its.feature#L172)?


Anyway, it feels that the discussion seems to have drifted.
Do you still have some concerns regarding stubbing the subject?

**Reply by DannyBen** [reply]
> Those links are all about the implicit block expectation syntax, not specifically one-liner expectation syntax.

How else can `is_expected` be used other than in `it { is_expected }`?
I never saw it in any other appearance.

> Was it?

I must have misread this:

> `its` for RSpec 3 extracted from rspec-core 2.x

as a sign that it is extracted since the core team do not want to mess with it, but still want to provide this functionality to anyone who needs it. It feels like `shoulda` matchers that are now replaced with expectations. I may be wrong, but I definitely got the feeling `its` is just something the rspec team agreed to create, but not encourage.

Add to the fact that the [one liner syntax page](https://relishapp.com/rspec/rspec-core/docs/subject/one-liner-syntax) does not mention `its`, even though it clearly belongs there and you get the overall "discouraged" smell. At least I do.

> Anyway, it feels that the discussion seems to have drifted.
> Do you still have some concerns regarding stubbing the subject?

Just enjoying the discussion. No, I feel I understand the reasoning enough to make an educated choice as to disabling this and other cops when I need to.

Thanks.

**Reply by pirj** [reply]
>  implicit block expectation syntax

There's some confusion here.

`is_expected` is defined as `expect(subject).to ...`. It is a **value** expectation syntax. They can be used interchangeably, and there are no arguments against `is_expected`. There are, though, arguments against using a literal `subject`. Subject should be named, and not referenced directly as `subject`. `expect(stock).to` is more intent-revealing than `expect(subject)`, as it doesn't require to look up what the `subject` is from the reader.
I'd say `is_expected` is preferred for one-liner expectations, where `subject` is defined somewhere close to those grouped one-liners.

In contrary, the **block** expectation syntax is `expect { some_method }.to ...`.

Those two syntaxes are different in a sense that the former is testing the returned value, while the latter is testing side effects.
Matchers usually only support one of them. E.g. `expect { puts "hi" }.to eq(true)` doesn't make sense.

Theoretically, `expect(stock).to change(rate).to(1)` is not supposed to work.
But it did, if `stock` was defined as a lambda, `let(:stock) { -> { rate.increment! } }`. This is what we call the implicit block expectation syntax. Something that doesn't look like a block but behaves like one.
It is now deprecated and will cease to work in RSpec 4.

`is_expected` will remain intact.

**Comment by pirj** [comment]
> its for RSpec 3 extracted from rspec-core 2.x
> as a sign that it is extracted since the core team do not want to mess with it

It's a personal preference, but I don't use `its`, as there are other ways to describe the same behaviour:
```diff
subject(:number) { 3 }

its(:odd?) { is_expected.to be true }
it { expect(number).to be_odd }
it { is_expected.to be_odd }
```

There are even cases where I can't tell if `its` has a direct substitute:
```ruby
it { expect(number.allbits?(1)).to be true }
it { is_expected.to be_allbits(1) }
```

**Reply by DannyBen** [reply]
Thanks for both messages. I didn't see anybody use the one liner syntax like this:

```ruby
it { expect(number).to be_odd }
```

it does not read right. I guess this is the place to use `specify`?

```ruby
specify { expect(number).to be_odd }
```

**Reply by pirj** [reply]
Fair enough. I don't have a definitive answer, and there's no written down agreement. [Related style guide issue](https://github.com/rubocop/rspec-style-guide/issues/64). Your example reads like a correct English sentence.
The only hint is from the [code defining `specify`](https://github.com/rspec/rspec-core/blob/6dede8ae1b9cce6e1de4baa912e77c97b587b446/lib/rspec/core/example_group.rb#L163):
```ruby
      # Useful for when your docstring does not read well off of `it`.
      # @example
      #  RSpec.describe MyClass do
      #    specify "#do_something is deprecated" do
      #      # ...
      #    end
      #  end
      define_example_method :specify
```
even though, personally, I'd use `it` here:
```ruby
      #  RSpec.describe MyClass do
      #    describe "#do_something" do
      #      it "is deprecated" do
      #        # ...
      #      end
      #    end
      #  end
```
and would reserve `specify` for examples with no implicit subject (`is_expected`) and no docstring, both multi-line and one-line.
```ruby
specify do
  expect(stock).to be_on_the_rise
end
it { is_expected.to be_on_the_rise }
specify { expect(input.split).to include('A', 'b') }
```

**Reply by DannyBen** [reply]
Yup. Seems right.
Thanks for the nice discussion and tips.

**Reply by pirj** [reply]
Likewise, @DannyBen !
